extmod/modbluetooth: Fixes for bluekitchen/btstack.

jimmo opened the pull request:
This PR brings btstack functionality up to match NimBLE (excluding L2CAP channels). This needed to be done anyway, but is high priority for BLE on the Pico W (see #10739).

The three main fixes are:
 - Make indicate/notify behave identically, and match NimBLE.
 - Include the value handle in the read/write complete events for GATT client. (Previously btstack always gave 65535 as the handle).
 - Make central-initiated MTU exchange work on btstack.

Also
 - Update tests to pass on both. ble_characteristic.py, ble_descriptor.py,ble_gap_advertise.py, ble_gap_connect.py, ble_gap_device_name.py, ble_gattc_discover_services.py, ble_gatt_data_transfer.py, ble_mtu.py,ble_subscribe.py now run identically on both.
 - Updates documentation (mainly to remove btstack-specific limitations).
 - Some minor NimBLE fixes, e.g. for #6864

TODO: More testing of btstack pairing/bonding.

_This work was funded through GitHub Sponsors._

jimmo committed:
extmod/modbluetooth: Merge gatts_notify/indicate implementation.

Makes gatts_notify and gatts_indicate work in the same way: by default they
send the DB value, but you can manually override the payload.

In other words, makes gatts_indicate work the same as gatts_notify.

Note: This removes support for queuing notifications and indications on
btstack when the ACL buffer is full. This functionality will be
reimplemented in a future commit.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/btstack: Fix indicate/notify queuing.

This adds a mechanism to track a pending notify/indicate operation that
is deferred due to the send buffer being full. This uses a tracked alloc
that is passed as the content arg to the callback.

This replaces the previous mechanism that did this via the global pending
op queue, shared with client read/write ops.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/btstack: Include value handle in client read/write events.

This replaces the previous pending operation queue (that used to also be
shared with pending server notify/indicate ops) with a single pending
operation per connection. This allows the value handle to be correctly
passed to the Python-level events.

Also re-structure GATT client event handling to simplify the packet handler
functions.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Use time.sleep_ms instead of time.sleep.

On unix, time.sleep is implemented as select(timeout=<time>) which means
that it does not run the poll hook during sleeping.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth/ble_mtu: Split peripheral/central-initiated.

btstack only supports central-initiated, so this allows us to have a test
that works on both (ble_mtu.py), and then another one for just the NimBLE
supported behavior (ble_mtu_peripheral.py).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/btstack: Fix MTU handling.

There was no event handler for central-initiated MTU exchange.

Fix truncation of notify/indicate to match NimBLE.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth/ble_subscribe: Use end_handle in desc discovery.

Obtaining the end_handle was added in cacc96d9.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/btstack: Add support for CCCD to allow client subscriptions.

This allows gatts_write(..., send_update=True) to work, which will send
notifications/indications to subscribed clients.

btstack already created the CCCD but writes to it were ignored.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Add test for descriptors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Fix flags for descriptor registration.

The att_flags for descriptors does not use the same bitfield as for
characteristics. This was leading to NimBLE descriptors getting the wrong
flags set.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth/ble_characteristic.py: Add write-no-response.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
examples/bluetooth/ble_temperature_central.py: Remove service tuple.

This is unused in the client, only needed in the server.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> Include the value handle in the read/write complete events for GATT client. (Previously btstack always gave 65535 as the handle).

This is important to support aioble, which requires this to work (which is why aioble only previously worked on NimBLE).

See also https://github.com/micropython/micropython-lib/pull/643 for a corresponding update to aioble related to other fixes.

jimmo committed:
tests/multi_bluetooth: Use multitest.broadcast instead of sleep.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Make all HCI transports trace in the same format.

 - Use HCI_TRACE macro consistently.
 - Use the same colour formatting.
 - Add a tool to convert to .pcap for Wireshark.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/btstack/modbluetooth_btstack: Add default services.

Even if the user doesn't call ble.gatts_register_services, always
provide the default services.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented on extmod/nimble/hal/hal_uart.c:
```diff
@@ -38,7 +38,10 @@
 #endif
 
-#define HCI_TRACE (0)
+#define HCI_TRACE (1)
```
should this stay disabled?

dpgeorge commented:
I reverted the commit that changed `run-multitests.py` to poll stdin during `wait()`, because it breaks the CI.  I'd prefer to fix that by making the unix port run scheduled code during `mp_hal_stdin_rx_chr()`.

dpgeorge commented on docs/library/bluetooth.rst:
```diff
@@ -514,19 +512,24 @@ writes from a client to a given characteristic, use
     set with :meth:`gatts_write <BLE.gatts_write>`) will be sent.
 
+    If *data* is not ``None``, then that value is sent to the client as part of
+    the notification. The local value will not be modified.
```
It looks like this paragraph was inadvertently moved here?  Because it no longer reads properly.

jimmo replied:
Yep. I've kept the order because i think it's clearer to go default first then override second, but fixed the wording.

dpgeorge commented on docs/library/bluetooth.rst:
```diff
@@ -514,19 +512,24 @@ writes from a client to a given characteristic, use
+    set with :meth:`gatts_write <BLE.gatts_write>`) will be sent.
+
+    If *data* is not ``None``, then that value is sent to the client as part of
+    the notification. The local value will not be modified.
```
As above, I think these paragraphs should be swapped.

dpgeorge commented on extmod/btstack/modbluetooth_btstack.c:
```diff
@@ -1109,90 +1069,152 @@ int mp_bluetooth_gatts_write(uint16_t value_handle, const uint8_t *value, size_t
     }
+    assert(err == ERROR_CODE_SUCCESS);
+    (void)err;
+    m_tracked_free(pending_op);
```
Should this free go outside `MICROPY_PY_BLUETOOTH_EXIT?

jimmo replied:
Yes. Fixed.

dpgeorge commented on tests/multi_bluetooth/ble_gap_advertise.py:
```diff
@@ -51,12 +54,12 @@ def irq(ev, data):
         pass
     ble.irq(irq)
-    ble.gap_scan(2 * ADV_TIME_S * 1000, 10000, 10000)
+    ble.gap_scan(2 * ADV_TIME_MS, 30000, 30000)
```
Was it intentional to change the 2nd and 3rd arguments here?

jimmo replied:
Yes, 30000 is the minimum.

dpgeorge commented on ports/unix/mpbthciport.c:
```diff
@@ -140,10 +144,10 @@ STATIC int configure_uart(void) {
 
     // 1Mbit (TODO: make this configurable).
-    speed_t brate = B1000000;
+    speed_t brate = B115200;
```
Why did this and the above CRTSCTS setting change?

jimmo replied:
That was accidentally left in from an experiment with the Pico W issues. Fixed.

dpgeorge commented:
I also pushed a commit to tidy up a comment.

dpgeorge commented on tools/hci_trace_to_pcap.py:
```diff
@@ -0,0 +1,129 @@
+            # These are always complete.
+            sys.stdout.buffer.write(
+                struct.pack(
+                    "!IIIII", timestamp, 0, len(data) + 4, len(data) + 4, _DIR_HOST_TO_CONTROLLER
```
I think the timestamp encoding should be:
```py
"!IIIII", timestamp // 1000, timestamp % 1000 * 1000, ...
```
Also the same change below.

jimmo replied:
Done

jimmo commented:
> I reverted the commit that changed `run-multitests.py` to poll stdin during `wait()`, because it breaks the CI. I'd prefer to fix that by making the unix port run scheduled code during `mp_hal_stdin_rx_chr()`.

OK. I've squashed out the commit and the revert.

Also rebased the other changes.

dpgeorge commented:
OK, thanks for updating, looks good now!

dpgeorge commented:
Merged!
